src/lib.rs

```diff
@@ -368,6 +391,135 @@ named!(parse_d_packet<&[u8], Option<u64>>,
+                     tag!("4") => { |_| ZType::AccessWatchpoint }));
+
+named!(parse_cond_or_command_expression<&[u8], Vec<u8>>,
+       do_parse!(tag!("X") >>
+                 len: hex_value >>
+                 tag!(",") >>
+                 expr: take!(len) >>
+                 (expr.to_vec())));
+
+named!(parse_condition_list<&[u8], Vec<Vec<u8>>>,
+       do_parse!(tag!(";") >>
+                 list: many1!(parse_cond_or_command_expression) >>
+                 (list)));
+
+fn maybe_condition_list<'a>(i: &'a[u8]) -> IResult<&'a [u8], Option<Vec<Vec<u8>>>> {
+    // An Incomplete here really means "not enough input to match a
+    // condition list", and that's OK.  An Error is *probably* that the
+    // input contains a command list rather than a condition list; the
+    // two are identical in their first character.  So just ignore that
+    // FIXME.
+    match parse_condition_list(i) {
+        Done(rest, v) => Done(rest, Some(v)),
+        Incomplete(_i) => Done(i, None),
+        Error(_) => Done(i, None),
+    }
+}
+
+named!(parse_command_list<&[u8], Vec<Vec<u8>>>,
+       // FIXME we drop the persistence flag here. 
+       do_parse!(tag!(";cmds") >>
+                 list: alt_complete!(do_parse!(persist_flag: hex_value >>
+                                               tag!(",") >>
+                                               cmd_list: many1!(parse_cond_or_command_expression) >>
+                                               (cmd_list)) |
+                                     many1!(parse_cond_or_command_expression)) >>
+                 (list)));
+
+fn maybe_command_list<'a>(i: &'a[u8]) -> IResult<&'a [u8], Option<Vec<Vec<u8>>>> {
+    // An Incomplete here really means "not enough input to match a
+    // command list", and that's OK.
+    match parse_command_list(i) {
+        Done(rest, v) => Done(rest, Some(v)),
+        Incomplete(_i) => Done(i, None),
+        Error(e) => Error(e),
+    }
+}
+
+named!(parse_cond_and_command_list<&[u8], (Option<Vec<Vec<u8>>>,
+                                           Option<Vec<Vec<u8>>>)>,
+       do_parse!(cond_list: maybe_condition_list >>
+                 cmd_list: maybe_command_list >>
+                 (cond_list, cmd_list)));
+
+fn parse_z_packet<'a>(i: &'a [u8]) -> IResult<&'a [u8], Command<'a>> {
+    let (rest, (action, type_, addr, kind)) = try_parse!(i, do_parse!(
+                                                      action: parse_z_action >>
+                                                      type_: parse_z_type >>
+                                                      tag!(",") >>
+                                                      addr: hex_value >>
+                                                      tag!(",") >>
+                                                      kind: hex_value >>
+                                                      (action, type_, addr, kind)));
+    println!("z packet remaining: {:?}", rest);
```

Comment:
Leftover debugging print.

Reply:
Doh, thank you.

Reply:
It might help with our sanity to just add the log crate and start sprinkling logging macros around.